### PR TITLE
Fix disjoint arguments to C-string kernels

### PR DESCRIPTION
**Pull request by connorjward**
I just want to make sure that I've interpreted your suggestions correctly. Tests are now passing.

@wence- please double check.

### CODE REVIEW

File: pyop2/codegen/rep2loopy.py [hunk 1]
```diff
@@ -655,12 +649,21 @@ def statement_functioncall(expr, context):
...
                                  predicates=predicates,
                                  id=id,
                                  depends_on=depends_on, depends_on_is_final=True)
 
+    # We cannot reconstruct the correct calling convention for C-string kernels
+    # without providing some additional context about the argument ordering.
+    # This is processed inside the ``emit_call_insn`` method of
+    # :class:`.PyOP2KernelCallable`.
+    if not hasattr(context, "function_args"):
+        context.function_args = {}
```
Comment (wence-) [inline pyop2/codegen/rep2loopy.py:663]: Just set this up in the `WrapperBuilder` constructor?

File: pyop2/codegen/rep2loopy.py [hunk 2]
```diff
@@ -655,12 +649,21 @@ def statement_functioncall(expr, context):
...
+    context.function_args[id] = function_args
```
Comment (wence-) [inline pyop2/codegen/rep2loopy.py:664]: Do we guarantee that the `id` is unique? I guess so, is a `CallInstruction` hashable, that would make me less nervous.

File: pyop2/codegen/rep2loopy.py [hunk 3]
```diff
@@ -204,11 +204,12 @@ class PyOP2KernelCallable(loopy.ScalarCallable):
...
-    fields = set(["name", "access", "arg_id_to_dtype", "arg_id_to_descr", "name_in_target"])
-    init_arg_names = ("name", "access", "arg_id_to_dtype", "arg_id_to_descr", "name_in_target")
+    fields = set(["name", "arguments", "access", "arg_id_to_dtype", "arg_id_to_descr", "name_in_target"])
+    init_arg_names = ("name", "arguments", "access", "arg_id_to_dtype", "arg_id_to_descr", "name_in_target")
 
-    def __init__(self, name, access, arg_id_to_dtype=None, arg_id_to_descr=None, name_in_target=None):
+    def __init__(self, name, arguments, access, arg_id_to_dtype=None, arg_id_to_descr=None, name_in_target=None):
         super(PyOP2KernelCallable, self).__init__(name, arg_id_to_dtype, arg_id_to_descr, name_in_target)
+        self.arguments = arguments
         self.access = access
```
Comment (wence-) [inline pyop2/codegen/rep2loopy.py:213]: I think that access is no longer used so we could remove it?

File: pyop2/codegen/rep2loopy.py [hunk 4]
```diff
@@ -562,7 +551,9 @@ def generate(builder, wrapper_name=None):
         wrapper = loopy.register_callable(
             wrapper,
             kernel.name,
-            PyOP2KernelCallable(name=kernel.name, access=tuple(builder.argument_accesses)))
+            PyOP2KernelCallable(name=kernel.name,
+                                arguments=context.function_args,
```
Comment (wence-) [inline pyop2/codegen/rep2loopy.py:555]: Ah, at this point we don't have the instruction available, so we pass the whole dict. Is the name a good one?

File: pyop2/codegen/rep2loopy.py (hunk 1 above)
Comment (connorjward) [reply pyop2/codegen/rep2loopy.py]: Now set up where `context` is defined.

File: pyop2/codegen/rep2loopy.py (hunk 2 above)
Comment (connorjward) [reply pyop2/codegen/rep2loopy.py]: `CallInstruction` objects aren't hashable. I've used the `name` attribute instead now.

File: pyop2/codegen/rep2loopy.py (hunk 4 above)
Comment (connorjward) [reply pyop2/codegen/rep2loopy.py]: If you mean "can we key the dict using `kernel.name`? then the answer is yes and I've done that.

File: pyop2/codegen/rep2loopy.py (hunk 2 above)
Comment (sv2518) [reply pyop2/codegen/rep2loopy.py]: Maybe it is possible to use the function here https://github.com/firedrakeproject/loopy/blob/94f191ce2545de5c959b19844d367c129473e80d/loopy/kernel/instruction.py#L441?

Comment (connorjward) [reply pyop2/codegen/rep2loopy.py]: Is the kernel `name` attribute not unique?

Comment (wence-) [reply pyop2/codegen/rep2loopy.py]: It must be, yes.

File: pyop2/codegen/rep2loopy.py (hunk 4 above)
Comment (wence-) [reply pyop2/codegen/rep2loopy.py]: I meant is `function_args` a good name for this dict?

Comment (connorjward) [reply pyop2/codegen/rep2loopy.py]: `kernel_parameters`?

Comment (wence-) [reply pyop2/codegen/rep2loopy.py]: Better, I think.

Comment (connorjward) [reply pyop2/codegen/rep2loopy.py]: Done.

**Review by wence- — APPROVED** [review]
Looks right, thanks.

**Pull request merged**